evalml/data_checks/invalid_target_data_check.py

```diff
@@ -209,7 +246,99 @@ def validate(self, X, y):
                     },
                 ).to_dict()
             )
+        return results
+
+    def _check_target_and_features_compatible(self, X, y, results):
+        if X is not None:
+            X = infer_feature_types(X)
+            X_index = list(X.index)
+            y_index = list(y.index)
+            X_length = len(X_index)
+            y_length = len(y_index)
+            if X_length != y_length:
+                DataCheck._add_message(
+                    DataCheckWarning(
+                        message="Input target and features have different lengths",
+                        data_check_name=self.name,
+                        message_code=DataCheckMessageCode.MISMATCHED_LENGTHS,
+                        details={
+                            "features_length": X_length,
+                            "target_length": y_length,
+                        },
+                    ),
+                    results,
+                )
+
+            if X_index != y_index:
+                if set(X_index) == set(y_index):
+                    DataCheck._add_message(
+                        DataCheckWarning(
+                            message="Input target and features have mismatched indices order",
+                            data_check_name=self.name,
+                            message_code=DataCheckMessageCode.MISMATCHED_INDICES_ORDER,
+                            details={},
+                        ),
+                        results,
+                    )
+                else:
+                    index_diff_not_in_X = list(set(y_index) - set(X_index))[:10]
+                    index_diff_not_in_y = list(set(X_index) - set(y_index))[:10]
```

Comment:
I like that you preempt the huge warning messages.  Are we able to have the message somehow communicate that these are only up to 10 of the mismatched indices?

Reply:
I'll add it to the message! :)